solutions/cf-1557B.mdx

```diff
@@ -0,0 +1,343 @@
+	<tbody>
+		<tr>
+			<td>$\texttt{arr}[i]$</td>
+			<td>6</td>
+			<td>3</td>
+			<td>4</td>
+			<td>2</td>
+			<td>1</td>
+		</tr>
+	</tbody>
+</table>
+</center>
+
+<center>
+<table>
+	<tbody>
+		<tr>
+			<th>Index:</th>
+			<th className="bg-yellow-100 dark:bg-yellow-800">0</th>
+			<th className="bg-green-100 dark:bg-green-800">1</th>
+			<th className="bg-green-100 dark:bg-green-800">2</th>
+			<th className="bg-red-100 dark:bg-red-800">3</th>
+			<th className="bg-blue-100 dark:bg-blue-800">4</th>
+		</tr>
+	</tbody>
+	<tbody>
+		<tr>
+			<td>$\texttt{arr}[i]$</td>
+			<td>6</td>
+			<td>3</td>
+			<td>4</td>
+			<td>2</td>
+			<td>1</td>
+		</tr>
+	</tbody>
+</table>
+</center>
+
+Thus, we need 4 subarrays at minimum.
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<Spoiler title="Time Complexity Proof">
+Sorting Requires $\mathcal{O}(N\log N)$, and repeated map insertions also require $\mathcal{O}(N\log N)$. Hence the overall time complexity is $\mathcal{O}(N\log N)$.
+</Spoiler>
+	
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+typedef long long ll;
+#define pb push_back
+#define FF first
+#define SS second
+ 
+void solve() {
+	int n, k; cin >> n >> k;
+	int arr[n];
+	int sarr[n];
+	for (int i = 0; i < n; i++) {
+		cin >> arr[i];
+		sarr[i] = arr[i];
+	}
+	map<int, int> next; // Map from one value to the value after it in the sorted array
+	sort(sarr, sarr + n);
+	for (int i = 0; i < n - 1; i++) {
+		next[sarr[i]]=sarr[i+1];
+	}
+	next[sarr[n-1]] = 1000000007; // Ridiculously Large Number
+	int ans = 1;
+	for (int i = 1; i < n; i++) {
+		// If the previous element in the unsorted array is not the same 
+		// as the previous element in the sorted array, we need at least 1 more segment
+		if (next[arr[i-1]]!=arr[i]) ans++; 
+	}
+	cout << (ans <= k ? "YES" : "NO") << endl;
+}
+ 
+int main() {
+	ios::sync_with_stdio(0); cin.tie(0);
+	int t; cin >> t; while (t--) solve();
+}
+```
+</CPPSection>
+<JavaSection>
+```java
+import java.io.*;
+import java.util.*;
+
+public class MoamenSubarrays {
+	public static void main (String[] args) throws IOException {
+		BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
+		StringTokenizer st = new StringTokenizer(f.readLine());
+	
+		int testNum = Integer.parseInt(st.nextToken());
+		for (int t = 0; t < testNum; t++) {
+			st = new StringTokenizer(f.readLine());
+			int n = Integer.parseInt(st.nextToken());
+			int k = Integer.parseInt(st.nextToken());
+			
+			int[] arr = new int[n];
+			int[] sarr = new int[n];
+	
+			st = new StringTokenizer(f.readLine());
+			for (int i = 0; i < n; i++) {
+				arr[i] = Integer.parseInt(st.nextToken());
+				sarr[i] = arr[i];
+			}
+			Arrays.sort(sarr);
+					      
+			TreeMap<Integer, Integer> next = new TreeMap<>(); // Map from one value to the value after it in the sorted array
+			for (int i = 0; i < n - 1; i++) {
+				next.put(sarr[i], sarr[i + 1]);
+			}
+			next.put(sarr[n - 1], (int) 1e9);
```

Comment:
again, if you're really against using Integer.MAX_VALUE, make a constant LARGE or something

Reply:
Changed everything to max value